### PR TITLE
WIP First draft for version field

### PR DESCRIPTION
**Pull request by cbuescher**
This is a first draft of a new field type aimed at storing software version strings like e.g. versions following the Semantic Versioning scheme. The field should offer exact matching and range queries similar to e.g. what we currently have for IP fields.
This POC contains of some early ideas about how we could encode the version strings into a binary representation that would allow ordering according to e.g. the Semantiv Versioning specification (i.e. precedence in part 11. in https://semver.org) but also potentially allow for more flexibility. Early tests for a "version" field type and a "version_range" range type show how basic search and aggs are possible with this encoding.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/VersionStringFieldMapper.java [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+    protected void doXContentBody(XContentBuilder builder, boolean includeDefaults, Params params) throws IOException {
+        super.doXContentBody(builder, includeDefaults, params);
+
+        if (nullValue != null) {
+            builder.field("null_value", nullValue);
+        }
+
+        if (includeDefaults || ignoreMalformed.explicit()) {
+            builder.field("ignore_malformed", ignoreMalformed.value());
+        }
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/VersionStringFieldMapper.java:378]: this must support all options that are supported by the parser, so it looks like some of them are missing

File: server/src/main/java/org/elasticsearch/index/mapper/VersionStringFieldMapper.java [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+        @Override
+        protected BytesRef indexedValueForSearch(Object value) {
+            if (value instanceof String) {
+                return encodeVersion((String) value, mode);
+            } else if (value instanceof BytesRef) {
+                // encoded string, need to re-encode
+                return encodeVersion(((BytesRef) value).utf8ToString(), mode);
+            } else {
+                throw new IllegalArgumentException("Illegal value type: " + value.getClass());
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/VersionStringFieldMapper.java:227]: maybe include the value as well in the error message?

File: server/src/main/java/org/elasticsearch/index/mapper/VersionEncoder.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        }
+
+        // pad all digit groups in main part with numeric marker and length bytes
+        BytesRefBuilder encodedVersion = new BytesRefBuilder();
+        prefixDigitGroupsWithLength(versionString, encodedVersion);
+
+        // encode whether version has pre-release parts
+        if (preReleaseId != null) {
+            encodedVersion.append(PRERELESE_SEPARATOR_BYTE);  // versions with pre-release part sort before ones without
+            String[] preReleaseParts = preReleaseId.substring(1).split(DOT_SEPARATOR_REGEX);
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/VersionEncoder.java:156]: If we want to support version strings that have consecutive dots, then I think we should use another split method as this one would skip the empty string in such a case.

File: server/src/main/java/org/elasticsearch/index/mapper/VersionEncoder.java [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+            boolean first = true;
+            for (String preReleasePart : preReleaseParts) {
+                if (first == false) {
+                    encodedVersion.append(DOT_SEPARATOR_BYTE);
+                }
+                boolean isNumeric = preReleasePart.chars().allMatch(x -> Character.isDigit(x));
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/VersionEncoder.java:162]: I wonder if we should treat the empty string as a numeric or as a string, does the spec say anything about this corner case?

**Review by jpountz — COMMENTED** [review]
I did a first quick round, this looks like a great start. The encoding makes sense to me.

File: server/src/main/java/org/apache/lucene/document/VersionRangeField.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    private static String toString(byte[] ranges, int dimension) {
+      byte[] min = new byte[BYTES];
+      System.arraycopy(ranges, 0, min, 0, BYTES);
+      byte[] max = new byte[BYTES];
+      System.arraycopy(ranges, BYTES, max, 0, BYTES);
+      return "["
+          + VersionEncoder.decodeVersion(new BytesRef(min), SortMode.SEMVER)
+          + " : "
+          + VersionEncoder.decodeVersion(new BytesRef(max), SortMode.SEMVER)
+          + "]";
```
[the remark this answers is not in the document]
Comment (cbuescher) [reply server/src/main/java/org/apache/lucene/document/VersionRangeField.java:125]: I wasn't sure either, the alternative is only to print the BytesRef hex output if we need some "toString".

File: server/src/main/java/org/elasticsearch/index/mapper/RangeType.java [hunk 6]
```diff
@@ -712,11 +895,29 @@ public int readLength(byte[] bytes, int offset) {
...
+         * Return the length of the value encoded at {@code offset} in {@code bytes}.
          */
         public abstract int readLength(byte[] bytes, int offset);
+
+        /**
+         * Return the number of positions the offset needs to be advances after reading the length
+         */
+        public int advanceBy() {
+            return 0;
+        };
```
[the remark this answers is not in the document]
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/mapper/RangeType.java:921]: BinaryRangeUtil#decodeRanges iterates over a BytesRef that contains both to/from. The other range types appartenly encode the length of the to/from byte field in the starting byte, for this new "FULL_BYTE" type we need to advance the pointer in the byte array by one. The other types don't advance the pointer. I'll add a comment along these lines.

File: server/src/main/java/org/elasticsearch/index/mapper/VersionEncoder.java [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+     * Encodes a version string given the given {@link SortMode}.
+     * First, the input version string is split into the following parts:
+     * <p>
+     * mainVersion(-preReleasePart)(+buildId)
+     *
+     */
+    public static BytesRef encodeVersion(String versionString, SortMode mode) {
+        //System.out.println("encoding: " + versionString);
+        if (legalVersionString(versionString) == false) {
+            throw new IllegalArgumentException("Illegal version string: " + versionString);
```
[the remark this answers is not in the document]
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/mapper/VersionEncoder.java:135]: Agreed, but that would require more detailed and potentially more costly checks than just checking against a regex. Validation is a whole sup-topic here I think.

File: server/src/main/java/org/elasticsearch/index/mapper/VersionEncoder.java (hunk 3 above)
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/mapper/VersionEncoder.java:156]: I haven't seen reason for supporting consecutive dots yet, currently I would try to catch that in initial validation. Do you have something in mind that would require this?

File: server/src/main/java/org/elasticsearch/index/mapper/VersionStringFieldMapper.java [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+            }
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName) {
+            failIfNoDocValues();
+            return new SortedSetOrdinalsIndexFieldData.Builder(CoreValuesSourceType.BYTES);
```
[the remark this answers is not in the document]
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/mapper/VersionStringFieldMapper.java:234]: Could you elaborate what you mean by this?

File: server/src/main/java/org/elasticsearch/index/mapper/VersionStringFieldMapper.java [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+            encodedVersion = encodeVersion(versionString, mode);
+        } catch (IllegalArgumentException e) {
+            if (ignoreMalformed.value()) {
+                context.addIgnoredField(name());
+                return;
+            } else {
+                throw e;
+            }
+        }
+        if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored())  {
```
[the remark this answers is not in the document]
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/mapper/VersionStringFieldMapper.java:340]: added this as TODO, some pointers would be helpful

File: server/src/main/java/org/elasticsearch/index/mapper/VersionStringFieldMapper.java (hunk 1 above)
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/mapper/VersionStringFieldMapper.java:378]: added output for "mode" for now

File: server/src/main/java/org/elasticsearch/index/mapper/VersionEncoder.java (hunk 4 above)
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/mapper/VersionEncoder.java:162]: The specs say that "Identifiers MUST NOT be empty". Identifiers are the dot-separated parts, so I think we can check this in some form of validation already (e.g. check and error on consecutive dots like mentioned above)

**Comment by cbuescher** [comment]
@jpountz thanks for the reviews and guidance, I opened https://github.com/elastic/elasticsearch/pull/59773 with the "version" field part of this draft, will follow up with a separate PR for the "version_range" work thats based on that, so I will close this draft.

**Pull request closed without merging**